From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangyueheng <1025067301@qq.com>
Date: Mon, 18 Jan 2021 22:08:24 +0800
Subject: [PATCH] Fix #119, DestroyBody may trigger EndContact.

Upstream-Status: Backport [https://github.com/pybox2d/pybox2d/commit/9172d95becefb40ba835639bb85b3707e7e94532]

Fixes a use-after-free when DestroyBody/DestroyJoint/DestroyFixture fire
callbacks that read the user data being released. PyPI's own cp310-cp313
2.3.10 wheels were built from a tree that already carries it.

---
 Box2D/Box2D_userdata.i | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/Box2D/Box2D_userdata.i b/Box2D/Box2D_userdata.i
index 79f9c1d..2fab9c8 100644
--- a/Box2D/Box2D_userdata.i
+++ b/Box2D/Box2D_userdata.i
@@ -43,21 +43,24 @@ public:
     }
 
     void DestroyBody(b2Body* body) {
-        Py_XDECREF((PyObject*)body->GetUserData());
+        void *userData = body->GetUserData();
         self->DestroyBody(body);
+        Py_XDECREF((PyObject*)userData);
     }
 
     void DestroyJoint(b2Joint* joint) {
-        Py_XDECREF((PyObject*)joint->GetUserData());
+        void *userData = joint->GetUserData();
         self->DestroyJoint(joint);
+        Py_XDECREF((PyObject*)userData);
     }
 }
 
 %extend b2Body {
 public:        
     void DestroyFixture(b2Fixture* fixture) {
-        Py_XDECREF((PyObject*)fixture->GetUserData());
+        void *userData = fixture->GetUserData();
         self->DestroyFixture(fixture);
+        Py_XDECREF((PyObject*)userData);
     }
     b2Fixture* __CreateFixture(b2FixtureDef* defn) {
         b2Fixture* ret;
-- 
2.43.0

